# Introduction

The goal of this notebook is to present the basic use of the "get" command of the Copernicus Marine Toolbox. You can find deeper information about the use of this command [in the documentation](https://toolbox-docs.marine.copernicus.eu/en/stable/python-interface.html#copernicusmarine.get).

## Import

We will need to import some specific packages to have this notebook running. They are grouped here.

In [2]:
from IPython.display import Video

In [3]:
import xarray as xr

## Installation

Normally the toolbox should be installed on the Jupyter Lab you are using. If it is not the case, you can install it via the following command:

In [4]:
!pip install copernicusmarine

And then you have to import it

In [5]:
import copernicusmarine

## Login

If you have already logged in a precedent step, this is not required. If not, you can use the following command and enter the username and password you have defined when you created your Copernicusmarine account

In [6]:
copernicusmarine.login()

INFO - 2026-09-29T12:19:31Z - Downloading Copernicus Marine data requires a Copernicus Marine username and password, sign up for free at: https://data.marine.copernicus.eu/register


Copernicus Marine username:

Abort: 

## Use of the get command

The get command is used to retrieve data from the data lake of Copernicus Marine. The data retrieved by get are different from those retrieved by the subset as they are the **original data** as delivered by the data producers

You can find the catalogue of all data at the [website of the Data Store](https://data.marine.copernicus.eu/products?disc=facets). There, you will need the **dataset_id** as well as some information about the name format to actually retrieve a file.  

### Getting information about a dataset

We will try to retrieve physical modelization of the Antarctic ocean and gather information about those using the get command. In the following video we will show how to find a dataset and find the structure of the data using the link provided above. 

In [7]:
Video("videos/Looking_for_data_info.webm", width=1200, html_attributes="controls")

Now that we have identified the dataset,we can use the **dry_run** option that gives the same output as the command, but without actually performing the downloading. It allows us to check the number of files and their sizes for example.

In [8]:
get_output_example = copernicusmarine.get(dataset_id="cmems_mod_arc_phy_anfc_6km_detided_P1D-m", dry_run=True)

INFO - 2026-09-29T12:19:53Z - Selected dataset version: "202311"
INFO - 2026-09-29T12:19:53Z - Selected dataset part: "default"
INFO - 2026-09-29T12:19:53Z - Listing files on remote server...
2it [00:00,  7.68it/s]


And if we inspect the output. It would be nice to right-click on the cell and select "Enable Scrolling for Outputs".

In [9]:
get_output_example

ResponseGet(files=[FileGet(s3_url='s3://mdl-native-10/native/ARCTIC_ANALYSISFORECAST_PHY_002_001/cmems_mod_arc_phy_anfc_6km_detided_P1D-m_202311/2021/07/20210705_dm-metno-MODEL-topaz5-ARC-b20210712-fv02.0.nc', https_url='https://s3.waw3-1.cloudferro.com/mdl-native-10/native/ARCTIC_ANALYSISFORECAST_PHY_002_001/cmems_mod_arc_phy_anfc_6km_detided_P1D-m_202311/2021/07/20210705_dm-metno-MODEL-topaz5-ARC-b20210712-fv02.0.nc', file_size=243.74546813964844, last_modified_datetime='2023-11-10T10:01:42.349000+00:00', etag='"fd98d5abed58abc2a421f36b4583868e-49"', file_format='.nc', output_directory=PosixPath('.'), filename='20210705_dm-metno-MODEL-topaz5-ARC-b20210712-fv02.0.nc', file_path=PosixPath('ARCTIC_ANALYSISFORECAST_PHY_002_001/cmems_mod_arc_phy_anfc_6km_detided_P1D-m_202311/2021/07/20210705_dm-metno-MODEL-topaz5-ARC-b20210712-fv02.0.nc'), file_status='DOWNLOADED'), FileGet(s3_url='s3://mdl-native-10/native/ARCTIC_ANALYSISFORECAST_PHY_002_001/cmems_mod_arc_phy_anfc_6km_detided_P1D-m_20231

As you can see, we see a long list of files, with informations about all of them. At the end are informations about the number of files and the total sizes of files to download. Let's inspect it.

In [10]:
print(f"The total number of files is {get_output_example.number_of_files_to_download} and the total size of download is {int(get_output_example.total_size/1000)} GB.")

The total number of files is 1923 and the total size of download is 470 GB.


We can also inspect information about just a file. Their information are in the list of the ResponseGet object and we can see what are the file names and their sizes

In [11]:
print(f"The file called '{get_output_example.files[12].filename}' has a size of {int(get_output_example.files[12].file_size)} MB")

The file called '20210717_dm-metno-MODEL-topaz5-ARC-b20210719-fv02.0.nc' has a size of 242 MB


### Selecting files

We see that downloading the entire dataset is out of reach for a teaching session. We should focus on just downloading one or two files.

For this, we can use the **filter** subcommand. It allows you to select files depending on a pattern found in their name. We will try to select the file "20210717_dm-metno-MODEL-topaz5-ARC-b20210719-fv02.0.nc" mentioned above with this. 

After a quick inspection of the other files, we see that most of the name if common but that there is one file per day for this dataset. However, there are two dates in the name: They can correspond to the start time and the end time of the data. It means that if we just select a date, there is likely another file whose name will also contain that date. Here, we can decide to choose the first date with the "_" as our pattern, as the second date is surrounded by "-". For the functioning of the filtering option, you have to surround your pattern with "*" which means that there could any string of characters before and after your pattern. For more explanation about this, you can look at [the "Standard Wildcard" section of this website](https://tldp.org/LDP/GNU-Linux-Tools-Summary/html/x11655.htm).

In [12]:
get_output_example_filter = copernicusmarine.get(dataset_id="cmems_mod_arc_phy_anfc_6km_detided_P1D-m", filter = "*20210717_*",dry_run=True)

INFO - 2026-09-29T12:20:15Z - Selected dataset version: "202311"
INFO - 2026-09-29T12:20:15Z - Selected dataset part: "default"
INFO - 2026-09-29T12:20:16Z - Listing files on remote server...
2it [00:00,  7.15it/s]


And we can inspect the ouput

In [13]:
get_output_example_filter

ResponseGet(files=[FileGet(s3_url='s3://mdl-native-10/native/ARCTIC_ANALYSISFORECAST_PHY_002_001/cmems_mod_arc_phy_anfc_6km_detided_P1D-m_202311/2021/07/20210717_dm-metno-MODEL-topaz5-ARC-b20210719-fv02.0.nc', https_url='https://s3.waw3-1.cloudferro.com/mdl-native-10/native/ARCTIC_ANALYSISFORECAST_PHY_002_001/cmems_mod_arc_phy_anfc_6km_detided_P1D-m_202311/2021/07/20210717_dm-metno-MODEL-topaz5-ARC-b20210719-fv02.0.nc', file_size=242.97493362426758, last_modified_datetime='2023-11-10T10:01:53.980000+00:00', etag='"a18aca96123045a3e707853e568e419d-49"', file_format='.nc', output_directory=PosixPath('.'), filename='20210717_dm-metno-MODEL-topaz5-ARC-b20210719-fv02.0.nc', file_path=PosixPath('ARCTIC_ANALYSISFORECAST_PHY_002_001/cmems_mod_arc_phy_anfc_6km_detided_P1D-m_202311/2021/07/20210717_dm-metno-MODEL-topaz5-ARC-b20210719-fv02.0.nc'), file_status='DOWNLOADED')], files_deleted=None, files_not_found=None, number_of_files_to_download=1, total_size=242.97493362426758, status=<StatusCode.

And see that we have one file of 242 MB

In [14]:
print(f"The total number of files is {get_output_example_filter.number_of_files_to_download} and the total size of download is {int(get_output_example_filter.total_size)} MB.")

The total number of files is 1 and the total size of download is 242 MB.


### Downloading and inspecting files

Now we can actually download and inspect one of the files. There are different types of files on the Copernicus Marine Data Store. The data we are interested in here are models outputs that are stored in the NetCDF format. You can find more information about this format on [its official website](https://www.unidata.ucar.edu/software/netcdf). For reading those data, we will need to use a package that is out of the toolbox, here [xarray in Python](https://xarray.dev/). 

First we actually download the file by removing the **dry_run** option and not storing the output in a variable, as we will not need it

In [15]:
copernicusmarine.get(dataset_id="cmems_mod_arc_phy_anfc_6km_detided_P1D-m", filter = "*20210717_*")

INFO - 2026-09-29T12:20:29Z - Selected dataset version: "202311"
INFO - 2026-09-29T12:20:29Z - Selected dataset part: "default"
INFO - 2026-09-29T12:20:29Z - Listing files on remote server...
2it [00:00,  7.78it/s]


ResponseGet(files=[FileGet(s3_url='s3://mdl-native-10/native/ARCTIC_ANALYSISFORECAST_PHY_002_001/cmems_mod_arc_phy_anfc_6km_detided_P1D-m_202311/2021/07/20210717_dm-metno-MODEL-topaz5-ARC-b20210719-fv02.0.nc', https_url='https://s3.waw3-1.cloudferro.com/mdl-native-10/native/ARCTIC_ANALYSISFORECAST_PHY_002_001/cmems_mod_arc_phy_anfc_6km_detided_P1D-m_202311/2021/07/20210717_dm-metno-MODEL-topaz5-ARC-b20210719-fv02.0.nc', file_size=242.97493362426758, last_modified_datetime='2023-11-10T10:01:53.980000+00:00', etag='"a18aca96123045a3e707853e568e419d-49"', file_format='.nc', output_directory=PosixPath('.'), filename='20210717_dm-metno-MODEL-topaz5-ARC-b20210719-fv02.0.nc', file_path=PosixPath('ARCTIC_ANALYSISFORECAST_PHY_002_001/cmems_mod_arc_phy_anfc_6km_detided_P1D-m_202311/2021/07/20210717_dm-metno-MODEL-topaz5-ARC-b20210719-fv02.0.nc'), file_status='DOWNLOADED')], files_deleted=None, files_not_found=None, number_of_files_to_download=1, total_size=242.97493362426758, status='000', messa

Now we can open the file with xarray to see what it contains. You can see that the file is in a succession of directories, as the get command copies the directory organization of the datalake. As a small tip, the path is indicated is the field **file_path**

In [16]:
physical_data = xr.open_dataset("ARCTIC_ANALYSISFORECAST_PHY_002_001/cmems_mod_arc_phy_anfc_6km_detided_P1D-m_202311/2021/07/20210717_dm-metno-MODEL-topaz5-ARC-b20210719-fv02.0.nc")

In [17]:
physical_data

<xarray.Dataset> Size: 1GB
Dimensions:        (y: 1137, x: 1185, time: 1, depth: 40)
Coordinates:
  * y              (y) float32 5kB -43.0 -42.94 -42.88 ... 27.88 27.94 28.0
  * x              (x) float32 5kB -36.0 -35.94 -35.88 ... 37.88 37.94 38.0
    longitude      (y, x) float32 5MB ...
    latitude       (y, x) float32 5MB ...
  * time           (time) datetime64[ns] 8B 2021-07-17T12:00:00
  * depth          (depth) float32 160B 0.0 2.0 4.0 6.0 ... 3e+03 3.5e+03 4e+03
Data variables: (12/19)
    stereographic  int32 4B ...
    model_depth    (y, x) float32 5MB ...
    so             (time, depth, y, x) float32 216MB ...
    thetao         (time, depth, y, x) float32 216MB ...
    vxo            (time, depth, y, x) float32 216MB ...
    vyo            (time, depth, y, x) float32 216MB ...
    ...             ...
    vxsi           (time, y, x) float32 5MB ...
    vysi           (time, y, x) float32 5MB ...
    sisnthick      (time, y, x) float32 5MB ...
    siage          (time, y, x) float32 5MB ...
    siconc_fy      (time, y, x) float32 5MB ...
    sialb          (time, y, x) float32 5MB ...
Attributes:
    institution:    Met Norway, Henrik Mohns plass 1, 0313 Oslo,  Norway
    references:     https://marine.copernicus.eu/
    source:         NERSC-HYCOM model fields
    contact:        <servicedesk.cmems@mercator-ocean.eu>
    credit:         E.U. Copernicus Marine Service Information (CMEMS)
    conventions:    CF-1.4
    field_date:     2021-07-17T12:00:00Z
    bulletin_date:  2021-07-19
    title:          Arctic Ocean Physics Analysis and Forecast, 6.25 km daily...
    bulletin_type:  Hindcast

### Exercise

Now it would be time for an exercise. Try and do the same thing as we have done before, but for downloading the biogeochemial modelisation data of the **Arctic in the daily format for the 10/05/2023**.

### Solution

The dataset is the **cmems_mod_arc_bgc_anfc_ecosmo_P1D-m** that you can find on [this page](https://data.marine.copernicus.eu/product/ARCTIC_ANALYSISFORECAST_BGC_002_004/services)

We can first check the pattern to download only file with the dry_run command:

In [18]:
example_bgc_data_dry_run = copernicusmarine.get(dataset_id="cmems_mod_arc_bgc_anfc_ecosmo_P1D-m", filter = "*20230510_*",dry_run=True)

INFO - 2026-09-29T12:20:43Z - Selected dataset version: "202105"
INFO - 2026-09-29T12:20:43Z - Selected dataset part: "default"
INFO - 2026-09-29T12:20:43Z - Listing files on remote server...
3it [00:00,  6.78it/s]


In [19]:
example_bgc_data_dry_run

ResponseGet(files=[FileGet(s3_url='s3://mdl-native-10/native/ARCTIC_ANALYSISFORECAST_BGC_002_004/cmems_mod_arc_bgc_anfc_ecosmo_P1D-m_202105/2023/05/20230510_dm-metno-MODEL-topaz5_ecosmo-ARC-b20230511-fv02.0.nc', https_url='https://s3.waw3-1.cloudferro.com/mdl-native-10/native/ARCTIC_ANALYSISFORECAST_BGC_002_004/cmems_mod_arc_bgc_anfc_ecosmo_P1D-m_202105/2023/05/20230510_dm-metno-MODEL-topaz5_ecosmo-ARC-b20230511-fv02.0.nc', file_size=579.0189352035522, last_modified_datetime='2023-11-10T16:36:08.413000+00:00', etag='"a5ac673897b56c6268e0c6280aadbb4c-116"', file_format='.nc', output_directory=PosixPath('.'), filename='20230510_dm-metno-MODEL-topaz5_ecosmo-ARC-b20230511-fv02.0.nc', file_path=PosixPath('ARCTIC_ANALYSISFORECAST_BGC_002_004/cmems_mod_arc_bgc_anfc_ecosmo_P1D-m_202105/2023/05/20230510_dm-metno-MODEL-topaz5_ecosmo-ARC-b20230511-fv02.0.nc'), file_status='DOWNLOADED')], files_deleted=None, files_not_found=None, number_of_files_to_download=1, total_size=579.0189352035522, status=

And since there is a file, we can now downloading it. 

In [20]:
example_bgc_data_dry_run = copernicusmarine.get(dataset_id="cmems_mod_arc_bgc_anfc_ecosmo_P1D-m", filter = "*20230510_*")

INFO - 2026-09-29T12:20:50Z - Selected dataset version: "202105"
INFO - 2026-09-29T12:20:50Z - Selected dataset part: "default"
INFO - 2026-09-29T12:20:50Z - Listing files on remote server...
3it [00:00,  7.77it/s]


In [21]:
bgc_data = xr.open_dataset("ARCTIC_ANALYSISFORECAST_BGC_002_004/cmems_mod_arc_bgc_anfc_ecosmo_P1D-m_202105/2023/05/20230510_dm-metno-MODEL-topaz5_ecosmo-ARC-b20230511-fv02.0.nc")

In [38]:
bgc_data

<xarray.Dataset> Size: 3GB
Dimensions:        (y: 1137, x: 1185, time: 1, depth: 40)
Coordinates:
  * y              (y) float32 5kB -43.0 -42.94 -42.88 ... 27.88 27.94 28.0
  * x              (x) float32 5kB -36.0 -35.94 -35.88 ... 37.88 37.94 38.0
    longitude      (y, x) float32 5MB ...
    latitude       (y, x) float32 5MB ...
  * time           (time) datetime64[ns] 8B 2023-05-10T12:00:00
  * depth          (depth) float32 160B 0.0 2.0 3.0 4.0 ... 3e+03 3.5e+03 4e+03
Data variables: (12/15)
    stereographic  int32 4B ...
    model_depth    (y, x) float32 5MB ...
    nppv           (time, depth, y, x) float32 216MB ...
    chl            (time, depth, y, x) float32 216MB ...
    kd             (time, depth, y, x) float32 216MB ...
    no3            (time, depth, y, x) float32 216MB ...
    ...             ...
    o2             (time, depth, y, x) float32 216MB ...
    si             (time, depth, y, x) float32 216MB ...
    expc           (time, depth, y, x) float32 216MB ...
    ph             (time, depth, y, x) float32 216MB ...
    dissic         (time, depth, y, x) float32 216MB ...
    spco2          (time, depth, y, x) float32 216MB ...
Attributes:
    institution:    NERSC, Thormoehlens gate 47, N-5006 Bergen, Norway
    history:        20230511:Created by program hyc2proj, version V0.3
    source:         NERSC-HYCOM model fields
    references:     http://topaz.nersc.no
    field_type:     Files based on file type archm
    Conventions:    CF-1.4
    field_date:     2023-05-10T12:00:00Z
    bulletin_date:  2023-05-11
    title:          Arctic Ocean Biogeochemistry Analysis and Forecast, 6.25 ...
    bulletin_type:  Hindcast

## Other types of data

So far we have seen examples of gridded data, but in the polar regions we also have satellite and in-situ data at our disposal. We will try to retrieve data for each of those two cases to see the difference

### Satellite data

You can find all the satellite datasets related to the Arctic Ocean [here](https://data.marine.copernicus.eu/products?disc=facets&facets=areas%7EArctic+Ocean--sources%7ESatellite+observations) and for the [Antarctic here](https://data.marine.copernicus.eu/products?disc=facets&facets=areas%7EAntarctic+Ocean--sources%7ESatellite+observations)

We can decide to go with the dataset for ocean colour for plankton. It is shown in the next video

In [22]:
Video("videos/Satellite_data_plankton.webm", width=1200, html_attributes="controls")

There are other datasets but they only contain data for the last months or few weeks before beind included in other datasets. We will see what is in the dataset with the --dry-run option

In [23]:
example_satellite_plankton = copernicusmarine.get(dataset_id="cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M",dry_run=True)

INFO - 2026-09-29T12:21:09Z - Selected dataset version: "202311"
INFO - 2026-09-29T12:21:09Z - Selected dataset part: "default"
INFO - 2026-09-29T12:21:09Z - Listing files on remote server...
1it [00:00, 22.00it/s]


In [24]:
example_satellite_plankton

ResponseGet(files=[FileGet(s3_url='s3://mdl-native-15/native/OCEANCOLOUR_ARC_BGC_L4_MY_009_124/cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M_202311/1997/19970901-19970930_cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M.nc', https_url='https://s3.waw3-1.cloudferro.com/mdl-native-15/native/OCEANCOLOUR_ARC_BGC_L4_MY_009_124/cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M_202311/1997/19970901-19970930_cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M.nc', file_size=3.494412422180176, last_modified_datetime='2025-06-26T13:33:44.880000+00:00', etag='"0fff38d5f32ad1a82f4cb94b364e7415"', file_format='.nc', output_directory=PosixPath('.'), filename='19970901-19970930_cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M.nc', file_path=PosixPath('OCEANCOLOUR_ARC_BGC_L4_MY_009_124/cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M_202311/1997/19970901-19970930_cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M.nc'), file_status='DOWNLOADED'), FileGet(s3_url='s3://mdl-native-15/native/OCEANCOLOU

In [25]:
print(f"The total number of files is {example_satellite_plankton.number_of_files_to_download} and the total size of download is {int(example_satellite_plankton.total_size)} MB.")

The total number of files is 340 and the total size of download is 1507 MB.


We see in the file names that the files correspond to a month of data, with date patterns such as "20240401-20240430". We will choose the month of July 2024, it should contain the pattern **20240701-20240731**.

In [26]:
copernicusmarine.get(dataset_id="cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M",filter="*20240701-20240731*")

INFO - 2026-09-29T12:21:23Z - Selected dataset version: "202311"
INFO - 2026-09-29T12:21:23Z - Selected dataset part: "default"
INFO - 2026-09-29T12:21:23Z - Listing files on remote server...
1it [00:00, 20.99it/s]


ResponseGet(files=[FileGet(s3_url='s3://mdl-native-15/native/OCEANCOLOUR_ARC_BGC_L4_MY_009_124/cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M_202311/2024/20240701-20240731_cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M.nc', https_url='https://s3.waw3-1.cloudferro.com/mdl-native-15/native/OCEANCOLOUR_ARC_BGC_L4_MY_009_124/cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M_202311/2024/20240701-20240731_cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M.nc', file_size=6.554759979248047, last_modified_datetime='2025-06-26T13:42:05.060000+00:00', etag='"453d78a20481f791415e34fbca2c0082"', file_format='.nc', output_directory=PosixPath('.'), filename='20240701-20240731_cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M.nc', file_path=PosixPath('OCEANCOLOUR_ARC_BGC_L4_MY_009_124/cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M_202311/2024/20240701-20240731_cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M.nc'), file_status='DOWNLOADED')], files_deleted=None, files_not_found=None, number_of

And now we can open the file to inspect it

In [27]:
plankton_satellite_data = xr.open_dataset('OCEANCOLOUR_ARC_BGC_L4_MY_009_124/cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M_202311/2024/20240701-20240731_cmems_obs-oc_arc_bgc-plankton_my_l4-multi-4km_P1M.nc')

In [28]:
plankton_satellite_data

<xarray.Dataset> Size: 38MB
Dimensions:        (time: 1, y: 1375, x: 1375)
Coordinates:
  * time           (time) datetime64[ns] 8B 2024-07-01
  * y              (y) float32 6kB -2.752e+06 -2.748e+06 ... 2.743e+06 2.748e+06
  * x              (x) float32 6kB -2.752e+06 -2.748e+06 ... 2.743e+06 2.748e+06
    lat            (y, x) float32 8MB ...
    lon            (y, x) float32 8MB ...
Data variables:
    stereographic  int32 4B ...
    CHL            (time, y, x) float32 8MB ...
    CHL_count      (time, y, x) float32 8MB ...
    CHL_error      (time, y, x) float32 8MB ...
Attributes: (12/33)
    netcdf_version:          v4
    Conventions:             CF-1.4
    platform:                Orbview-2/SeaStar,Aqua,ENVISAT,JPSS-1,Sentinel-3...
    product_level:           L4
    contact:                 technical@gos.artov.isac.cnr.it
    institution:             CNR-GOS
    ...                      ...
    references:              Zoffoli, M.L. et al. (2025) CIAO: A machine-lear...
    product_version:         v202506
    comment:                 Chl-a concentration retrieved from OC-CCI v.6 us...
    stop_date:               2024-07-31
    creation_date:           2025-04-12
    creation_time:           13:30:16 UTC

### In-situ data

There are also some in-situ data, we will show in the next video

In [29]:
Video("videos/Selection_in_situ_data.webm", width=1200, html_attributes="controls")

Here the data are ordered by platform instead of dates. To find the data from the polar zones, we have to identify platforms that are located there and then download them. From the video we have found a platform in the polar cicle and we can download its data